src/subclass/cell_renderer.rs

```diff
@@ -0,0 +1,658 @@
+    }
+    if !natptr.is_null() {
+        *natptr = nat_size;
+    }
+}
+
+unsafe extern "C" fn cell_renderer_get_preferred_width_for_height<T: ObjectSubclass>(
+    ptr: *mut gtk_sys::GtkCellRenderer,
+    wdgtptr: *mut gtk_sys::GtkWidget,
+    height: c_int,
+    min_width_ptr: *mut c_int,
+    nat_width_ptr: *mut c_int,
+) where
+    T: CellRendererImpl,
+{
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: CellRenderer = from_glib_borrow(ptr);
+    let widget: Widget = from_glib_borrow(wdgtptr);
+
+    let (min_width, nat_width) = imp.get_preferred_width_for_height(&wrap, &widget, height);
+    if !min_width_ptr.is_null() {
+        *min_width_ptr = min_width;
+    }
+    if !nat_width_ptr.is_null() {
+        *nat_width_ptr = nat_width;
+    }
+}
+
+unsafe extern "C" fn cell_renderer_get_aligned_area<T: ObjectSubclass>(
+    ptr: *mut gtk_sys::GtkCellRenderer,
+    wdgtptr: *mut gtk_sys::GtkWidget,
+    flags: gtk_sys::GtkCellRendererState,
+    cellarea: *const gdk_sys::GdkRectangle,
+    alignedptr: *mut gdk_sys::GdkRectangle,
+) where
+    T: CellRendererImpl,
+{
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: CellRenderer = from_glib_borrow(ptr);
+    let widget: Widget = from_glib_borrow(wdgtptr);
+
+    let rectangle = imp.get_aligned_area(
+        &wrap,
+        &widget,
+        from_glib(flags),
+        &from_glib_borrow(cellarea),
+    );
+    if !alignedptr.is_null() {
+        *alignedptr = *rectangle.to_glib_none().0;
+    }
+}
+
+unsafe extern "C" fn cell_renderer_render<T: ObjectSubclass>(
+    ptr: *mut gtk_sys::GtkCellRenderer,
+    crptr: *mut cairo_sys::cairo_t,
+    wdgtptr: *mut gtk_sys::GtkWidget,
+    bgptr: *const gdk_sys::GdkRectangle,
+    cellptr: *const gdk_sys::GdkRectangle,
+    flags: gtk_sys::GtkCellRendererState,
+) where
+    T: CellRendererImpl,
+{
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: CellRenderer = from_glib_borrow(ptr);
+    let widget: Widget = from_glib_borrow(wdgtptr);
+    let cr: cairo::Context = from_glib_borrow(crptr);
+
+    imp.render(
+        &wrap,
+        &cr,
+        &widget,
+        &from_glib_borrow(bgptr),
+        &from_glib_borrow(cellptr),
+        from_glib(flags),
+    );
+}
+
+unsafe extern "C" fn cell_renderer_activate<T: ObjectSubclass>(
+    ptr: *mut gtk_sys::GtkCellRenderer,
+    evtptr: *mut gdk_sys::GdkEvent,
+    wdgtptr: *mut gtk_sys::GtkWidget,
+    pathptr: *const c_char,
+    bgptr: *const gdk_sys::GdkRectangle,
+    cellptr: *const gdk_sys::GdkRectangle,
+    flags: gtk_sys::GtkCellRendererState,
+) -> glib_sys::gboolean
+where
+    T: CellRendererImpl,
+{
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: CellRenderer = from_glib_borrow(ptr);
+    let widget: Widget = from_glib_borrow(wdgtptr);
+    let evt: Option<gdk::Event> = from_glib_none(evtptr);
```

Comment:
```suggestion
    let evt: Option<gdk::Event> = from_glib_borrow(evtptr);
```

And then below `evt.as_ref()`

Reply:
Actually `let evt: gdk::Event = from_glib_borrow(evtptr);` works here, and then pass `&evt`

Reply:
Is it guaranteed to be never NULL? Then this is fine. I thought you might have had a reason for making it an `Option` :)

Reply:
the generated activate expect a reference to the GdkEvent so I did the same.

Reply:
:+1: